src/EntityFramework.SqlServer/SqlGen/SqlGenerator.cs

```diff
@@ -2639,6 +2648,36 @@ public override ISqlFragment Visit(DbSkipExpression e)
             return result;
         }
 
+        // We need to order by row_number unless ordering list of entities and we are sorting based on entity key.
+        // In case of composite key, all parts of the key must be used.
+        private bool NeedsOrderByRowNumber(IList<DbSortClause> sortClauses)
+        {
+            var propertyExpressions = sortClauses
+                .Where(c => c.Expression is DbPropertyExpression)
+                .Select(c => (DbPropertyExpression)c.Expression)
+                .ToList();
+
+            var propertyInstanceTypes = propertyExpressions
+                .Select(c => c.Instance.ResultType.EdmType)
+                .ToList();
+
+            var distinctEntityTypes = propertyInstanceTypes.OfType<EntityType>().Distinct().ToList();
+            if (distinctEntityTypes.Count != 1)
+            {
+                return true;
+            }
+
+            foreach (var keyProperty in distinctEntityTypes.First().KeyProperties)
+            {
+                if (propertyExpressions.All(e => e.Property != keyProperty))
+                {
+                    return true;
```

Comment:
I wonder not emitting row_number() if things are ordered on the keys is worth the effort. Looks a bit involved. Thoughts?